text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1022 @@
+managers in Glimmer and Ember may change to keep them compatible, but they will
+not make major changes without first getting community input, and will be
+considered part of the public API of Ember.
+
+## Prior Art
+
+Part of the challenge in the original Angle Bracket components RFC was
+attempting to design without implementation, testing, usage, and feedback.
+Glimmer.js provided an early method to experiment, but because it was not widely
+adopted there wasn't much feedback from larger-scale usage. This in part
+motivated the [component manager
+RFC](https://github.com/emberjs/rfcs/blob/master/text/0213-custom-components.md),
+which enabled experimentation in Ember directly, and set us up for having
+multiple implementations of component APIs which were interchangeable.
+
+As such, we now have two reference implementations which can be referred to:
+
+* [Glimmer.js](https://glimmerjs.com/), the framework that this component API is
+  based on, and will be cross-compatible with.
+* [sparkles-component](https://github.com/rwjblue/sparkles-component), an an
+  implementation of the Glimmer.js component API using Ember's component
+  managers. It is usable in Ember today.
+
+Both of these have minor differences from the API proposed in this RFC, mainly
+because they were made before the [element modifiers
+RFC](https://github.com/emberjs/rfcs/blob/master/text/0373-Element-Modifier-Managers.md)
+was accepted and opened up additional design possibilities. However, they serve
+as valuable data points for the viability of a simpler component API, and inform
+the design accordingly.
+
+## Detailed design
+
+Glimmer components have the following interface:
+
+```ts
+interface GlimmerComponentt {
+  args: unknown;
+
+  isDestroying: boolean;
+  isDestroyed: boolean;
+
+  constructor(owner: Owner, args: unknown): void;
+
+  didUpdateArgs(): void;
+  willDestroy(): void;
+}
+```
+
+In addition, two new element modifiers will fill in the gaps left by the
+`didInsertElement`, `didRender`, and `willDestroyElement` lifecycle hooks:
+
+* `{{did-render}}`
+* `{{will-destroy}}`
+
+### Constructor
+
+The constructor for Glimmer components receives two arguments: The owner
+instance and the named arguments object. Both of these arguments should
+conventionally be passed to `super` immediately, and then accessed through
+decorated service properties, `getOwner`, and `this.args` respectively:
+
+```js
+class PersonComponent extends GlimmerComponent {
+  @service profile;
+
+  constructor() {
+    super(...arguments);
+
+    let owner = getOwner(this);
+    let profileService = this.profile;
+
+    let firstName = this.args.firstName;
+  }
+}
+```
+
+These arguments are passed to the constructor so that they can be used for
+initial setup of the component. Since Glimmer components do not have an `init`
+hook, and they do not have `willRender`/`willInsertElement`, the constructor is
+only method for setting up the initial state of the component before the first
```

Comment:
This is true, but we should motivate that design so it doesn't seem like an arbitrary restriction.

Specifically, we want to encourage people to model initialization via getters (and tracked getters, if the value changes over time). This helps avoid expensive upfront initialization, and switches to a "pull-based" model of initialization that only does computation for values that are actually needed in the current rendering pass.